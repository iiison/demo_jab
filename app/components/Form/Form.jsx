import React, { createContext } from 'react'
import { PropTypes } from 'prop-types'

import { useSetState } from '$UTILS/reactUtils'
import validations from './validationRules'

export const FormCtx = createContext({})

export function useFormCTXProvider() {
  const [data, setData] = useSetState()

  return {
    data,
    setData,
    FormCTXProvider : FormCtx.Provider,
  }
}

function validateRule({ id, value, displayName, ruleValue, ruleArgs, validation }) {
  if (validation) {
    const stringifiedValue = value.toString()
    const isRuleSatisfied = ( ruleValue !== 'required' && !stringifiedValue )
      ? true
      : validation.rule.apply(null, ruleArgs).test(value.toString())

    let error = ''

    if (!isRuleSatisfied) {
      error = validation.formatter.apply(null, [displayName || id, ...ruleArgs])

      return error
    }
  } else {
    throw `invalid validation rule: ${ruleValue}, please use an existing validation rule name or pass a custom function with same name through 'customRules' prop in Input: ${id}. Rule value should be an object with keys: 'rule' as an Regex and 'formatter' as a function, that formats the value.` // eslint-disable-line
  }

  return ''
}

export function validateField({
  id,
  value,
  validate,
  displayName,
  customRules = {}
}) {
  let error = ''
  const rules = validate ? validate.split('|') : ''

  if (rules.length > 0) {
    for (const rule of rules) {
      if (!rule) {
        continue
      }

      const ruleDetails = rule.split('-')
      const [ruleValue, ...ruleArgs] = ruleDetails
      const validation = validations[ruleValue] || customRules[ruleValue]

      try {
        error = validateRule({ id, value, displayName, ruleValue, ruleArgs, validation })

        if (error !== '') {
          break
        }
      } catch (invalidRuleError) {
        throw invalidRuleError
      }
    }
  }

  return error
}

export function useUpdateField(setData, isNewField) {
  return (field, shouldRunValidation = true) => {
    const { id, isPristine, error } = field
    let validationError = error

    if (id === undefined) {
      throw new Error(`Please pass id to ${JSON.stringify(field)}`)
    }

    const value = isNewField === true && field.value === undefined ? '' : field.value

    if (isNewField !== true && isPristine === false && shouldRunValidation === true) {
      validationError = validateField(field)
    }

    setData({
      [id] : {
        ...field,
        value,
        error : validationError
      }
    })

  }
}

export default function Form({ children }) {
  const {
    data,
    setData,
    FormCTXProvider
  } = useFormCTXProvider()
  const addField = useUpdateField(setData, true)
  const updateField = useUpdateField(setData, false)
  const value = {
    data,
    addField,
    updateField,
    validateField
  }

  return (
    <FormCTXProvider value={value}>
      <form action="" className='col-12 grid'>
        {children}
      </form>
    </FormCTXProvider>
  )
}

Form.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
}

