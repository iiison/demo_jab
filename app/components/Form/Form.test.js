import React, { useContext } from 'react'
import { renderHook, act } from '@testing-library/react-hooks'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Form, {
  FormCtx,
  validateField,
  useUpdateField,
  useFormCTXProvider
} from './Form'

const TestComponent = () => {
  const { value } = useContext(FormCtx)

  return <div>value is: {value}</div>
}

describe('>> COMPONENT -- Form Tests', () => {
  describe('• HOOK - useFormCTXProvider Tests', () => {
    it('Check if `data` from returned values is an empty object', () => {
      const { result } = renderHook(() => useFormCTXProvider())
      const { data, setData } = result.current

      expect(data).toEqual({})
      expect(setData).toBeDefined();
    })

    it('Check if `setData` function sets value to the state `data`', () => {
      const { result } = renderHook(() => useFormCTXProvider())
      const { data, setData } = result.current

      expect(data).toEqual({})

      const newData = { a : 1 }

      act(() => setData(newData))

      expect(result.current.data).toEqual(newData)
    })

    it('Check if `FormCTXProvider` returns the right context value', () => {
      const { result } = renderHook(() => useFormCTXProvider())
      const { FormCTXProvider } = result.current
      const value = { value : 'value' }

      render(
        <FormCTXProvider value={value}>
          <TestComponent />
        </FormCTXProvider>
      )

      expect(screen.getByText(/^value is:/)).toHaveTextContent(`value is: ${value.value}`)
    })
  })

  describe('• HOOK - useUpdateField Tests', () => {
    const field = {
      id : 'test',
      value : 'some value'
    }
    it('Check if the hook returns a function', () => {
      const result = useUpdateField(() => {})

      expect(typeof result).toBe('function')
    })

    it('Check if the returned function calls the parent function setData function', () => {
      const setData = jest.fn()
      const updateField = useUpdateField(setData)

      updateField(field, false)

      expect(setData).toHaveBeenCalledWith({
        test : {
          ...field,
          error : undefined
        }
      })
    })

    it('Throws error if field doesn\'t have id property', () => {
      const fieldCopy = { ...field }

      delete fieldCopy.id

      const setData = jest.fn()
      const updateField = useUpdateField(setData)

      try {
        updateField(fieldCopy)
      } catch(err) {
        expect(err.message).toBe('Please pass id to {"value":"some value"}')
      }
    })

    it('sets field value to empty string if isNewField is true', () => {
      const fieldCopy = { ...field }

      delete fieldCopy.value

      const setData = jest.fn()
      const updateField = useUpdateField(setData, true)

      updateField(fieldCopy, false)

      expect(setData).toHaveBeenCalledWith({
        test : {
          ...field,
          error : undefined,
          value : ''
        }
      })
    })

    it('validates the field when isPristine is false', () => {
      const fieldCopy = { ...field }

      fieldCopy.value = 'some value'
      fieldCopy.isPristine = false
      fieldCopy.validate = 'required|email'

      const setData = jest.fn()
      const updateField = useUpdateField(setData, false)

      updateField(fieldCopy)
      expect(setData).toHaveBeenCalledWith({
        test : {
          ...fieldCopy,
          error : 'test is not valid email'
        }
      })
    })
  })

  describe('• Component Helper - validateField Tests', () => {
    const field = {
      id : 'test',
      value : 'some value'
    }

    it('throws error if invalid validation rule name is passed', () => {
      const fieldCopy = { ...field }
      
      fieldCopy.validate = 'wrong'

      try {
        validateField(fieldCopy)
      } catch (err) {
        expect(err).toBeDefined()
      }
    })

    it('returns empty string if no validation rule is passed', () => {
      const result = validateField(field)

      expect(result).toBe('')
    })

    it('returns empty string if validation rule passed validation', () => {
      const fieldCopy = { ...field }

      fieldCopy.validate = 'required'

      const result = validateField(fieldCopy)

      expect(result).toBe('')
    })

    it('returns error if validation rule fails', () => {
      const fieldCopy = { ...field }

      fieldCopy.value = 'email'

      const result = validateField(fieldCopy)

      expect(result).toBeDefined()
    })
  })

  describe('• Component - UI Tests', () => {
    it('renders the Form component snapshot', () => {
      const { container } = render(<Form><p>Nothing!</p></Form>)
      expect(container).toMatchSnapshot()
    })
  })
})

