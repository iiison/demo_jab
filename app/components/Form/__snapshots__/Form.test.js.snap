// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>> COMPONENT -- Form Tests â€¢ Component - UI Tests renders the Form component snapshot 1`] = `
<div>
  <form
    action=""
    class="col-12 grid"
  >
    <p>
      Nothing!
    </p>
  </form>
</div>
`;
